src/dune_cache/dune_cache.ml

```diff
@@ -285,23 +273,30 @@ module Cache = struct
                ; digest = effective_hash
                })
     in
-    let f () =
-      let+ promoted = Result.List.map ~f:promote paths in
-      let metadata_path = FSSchemeImpl.path (path_meta cache) key
-      and files = List.map ~f:file_of_promotion promoted in
-      let metadata_file : Metadata_file.t = { metadata; files } in
-      Path.mkdir_p (Path.parent_exn metadata_path);
-      Io.write_file metadata_path
-        (Csexp.to_string (Metadata_file.to_sexp metadata_file));
-      let f = function
-        | Already_promoted file when cache.duplication_mode <> Copy ->
-          cache.handler (Dedup file)
-        | _ -> ()
-      in
-      List.iter ~f promoted;
-      promoted
+    let+ promoted = Result.List.map ~f:promote paths in
+    let metadata_path = FSSchemeImpl.path (path_meta cache) key
+    and metadata_tmp_path = Path.relative cache.temp_dir "metadata"
+    and files = List.map ~f:file_of_promotion promoted in
+    let metadata_file : Metadata_file.t = { metadata; files } in
+    let metadata = Csexp.to_string (Metadata_file.to_sexp metadata_file) in
+    Io.write_file metadata_tmp_path metadata;
+    let () =
+      try
```

Comment:
It is usually preferable to put `try...with` blocks only around the expression that we expect to raise. Here, we could write this:

```ocaml
match Io.read_file metadata_path with
| s -> ...
| exception Sys_error _ -> ...
```

Which is slightly better as the exception handler won't trigger if it is `User_warning.emit` that raises. It's quite unlikely `User_waning.emit` would raise, but in theory it could since it prints to stdout. It also makes the intent a bit clearer IMO.